### PR TITLE
Validate trx

### LINKED ISSUE #36: Tron TransactionBuilder - Implement validateRawTransaction()
# Overview
The Tron `TransactionBuilder` class defines a method called [`validateRawTransaction()`](https://github.com/BitGo/bitgo-account-lib/blob/master/src/coin/trx/transactionBuilder.ts#L116-L118). The method currently performs zero validation on the raw transaction passed into it.

Let's implement the method to perform the necessary validation. Specifically: parse the transaction `raw_data_hex` and compare it with the `raw_data` to assert that they are the same.

# Definition of Done
- Implement Tron's `TransactionBuilder::validateRawTransaction()` method.
- Add test(s) to cover these changes.

### PR DESCRIPTION
**Pull request by leviathan-mist**
Closes #36 

Adding methodology for validation check for a raw transaction

### CODE REVIEW

File: src/coin/trx/transactionBuilder.ts [hunk 1]
```diff
@@ -117,12 +122,56 @@ export class TransactionBuilder extends BaseTransactionBuilder {
...
+   * transaction.
+   *
+   * The contents to be validated are
+   * 1. The transaction id
+   * 2. The expiration date
+   * 3. The timestamp
+   * 4. The contract
+   * @param rawTransaction The raw transaction to be validated
+   */
   validateRawTransaction(rawTransaction: any) {
```
Comment (dannybitgo) [inline src/coin/trx/transactionBuilder.ts]: you can update the input and return types of this to be like this:
```
validateRawTransaction(rawTransaction: TransactionReceipt | string) : void {
```

File: src/coin/trx/transactionBuilder.ts [hunk 2]
```diff
@@ -117,12 +122,56 @@ export class TransactionBuilder extends BaseTransactionBuilder {
...
-    // TODO: parse the transaction raw_data_hex and compare it with the raw_data
+    //TODO: Validation of signature
+    /** @inheritDoc Specifically, checks hex underlying transaction hashes to correct transaction ID. */
```
Comment (dannybitgo) [inline src/coin/trx/transactionBuilder.ts]: This inherit Doc doesn't belong here - it should be around line 174 above the `validateTransaction` method

**Comment by dannybitgo** [comment]
@leviathan-mist thanks again for your work here. feel free to ask any questions here if you get blocked or want to clarify anything

File: src/coin/trx/transactionBuilder.ts (hunk 1 above)
Comment (leviathan-mist) [reply src/coin/trx/transactionBuilder.ts]: Have added TransactionReceipt and void type to the method.

File: src/coin/trx/transactionBuilder.ts (hunk 2 above)
Comment (leviathan-mist) [reply src/coin/trx/transactionBuilder.ts]: Sure, have added the documentation to the correct lines. Thanks !!!

File: src/coin/trx/transactionBuilder.ts [hunk 3]
```diff
@@ -118,8 +123,60 @@ export class TransactionBuilder extends BaseTransactionBuilder {
...
+    let currTransaction: TransactionReceipt;
+    // rawTransaction can be either Stringified JSON OR
+    // it can be a regular JSON object (not stringified).
+    if (typeof rawTransaction === 'string') {
+      try {
+        currTransaction = JSON.parse(rawTransaction);
+      } catch (e) {
+        throw new ParseTransactionError('There was error in parsing the JSON string');
+      }
+    } else {
```
Comment (dannybitgo) [inline src/coin/trx/transactionBuilder.ts]: hey I'd add one more check here:
```
else if (_.isObject(rawTransaction)) {
  currTransaction = rawTransaction;
} else {
  // throw some error saying its an entirely invalid input
}
```

File: test/unit/coin/trx/transactionBuilder.ts [hunk 4]
```diff
@@ -167,10 +170,35 @@ describe('Tron', function() {
...
+
+    it('should throw exception of invalid expiration time', () => {
+      const txJson = JSON.stringify(UnsignedInvalidExpirationBuildTransaction);
+      should.throws(() => txBuilder.from(txJson));
+    });
+
+    it('should throw exception of non-existence of contract', () => {
+      const txJson = JSON.stringify(UnsignedInvalidContractBuildTransaction);
+      should.throws(() => txBuilder.from(txJson));
     });
```
Comment (dannybitgo) [inline test/unit/coin/trx/transactionBuilder.ts]: Can you add two more test cases to test the success case:
```
 it('should validate JSON transaction', () => {
      const txJson = SomeValidTransaction;
      should.doesNotThrow(() => txBuilder.from(txJson));
    });	
 it('should validate stringified JSON transaction', () => {
      const txJsonString = JSON.stringify(SomeValidTransaction);
      should.throws(() => txBuilder.from(txJsonString));
    });	 
```

Comment (leviathan-mist) [reply test/unit/coin/trx/transactionBuilder.ts]: Cool !!! I have added those two in the tests

File: src/coin/trx/transactionBuilder.ts (hunk 3 above)
Comment (leviathan-mist) [reply src/coin/trx/transactionBuilder.ts]: Thanks !!! I have fixed it

File: src/coin/trx/transactionBuilder.ts [hunk 5]
```diff
@@ -116,8 +120,63 @@ export class TransactionBuilder extends BaseTransactionBuilder {
...
+    }
+    // Validate the timestamp from the raw-data-hex
+    if (currTransaction.raw_data.timestamp != decodedRawDataHex.timestamp) {
+      throw new InvalidTransactionError('Transaction has not have a valid timetamp');
+    }
+    // Transaction contract must exist
+    if (!currTransaction.raw_data.contract) {
+      throw new InvalidTransactionError('Transaction contracts are empty');
+    }
+    function _isObject(rawTransaction: TransactionReceipt) {
```
Comment (dannybitgo) [inline src/coin/trx/transactionBuilder.ts]: this function isn't getting called

Comment (dannybitgo) [reply src/coin/trx/transactionBuilder.ts]: also i apologize if it wasn't clear, but i was suggesting an additional check on line 149, using lodash's `isObject` function.
More specifically, please replace lines 149-151 with this exact code:
```
else if (_.isObject(rawTransaction)) {
  currTransaction = rawTransaction;
} else {
  throw new InvalidTransactionError('Transaction is not an object or stringified json');
}
```
and then remove lines 177-179

Comment (dannybitgo) [reply src/coin/trx/transactionBuilder.ts]: And at the top of the file you'll have to add `import * as _ from 'lodash'`
Then you can use the `_.isObject` function that the lodash library provides

Comment (leviathan-mist) [reply src/coin/trx/transactionBuilder.ts]: Actually you were clear about it. And I had already added the object check using inbuilt function. But during the three way merge, it got removed. I have fixed it now using the lodash lib. 

thanks !!!

**Pull request merged**